File: modules/shared/utils/grpc/src/base_types.cpp
#include "oink_judge/utils/grpc/base_types.h"

#include "oink_judge/utils/grpc/auth_metadata_plugin.h"
#include "oink_judge/utils/grpc/auth_server_interceptor.h"
#include "oink_judge/utils/grpc/config_utils.h"
#include "oink_judge/utils/grpc/factories.hpp"

#include <oink_judge/config/common_utils.h>
#include <oink_judge/factory/parameterized_type_factory.hpp>
#include <oink_judge/utils/filesystem.h>

#include <grpcpp/grpcpp.h>
#include <grpcpp/security/credentials.h>

#include <string>
#include <vector>

namespace oink_judge::utils::grpc {

auto registerChannelType() -> void {
    ChannelFactory::instance().registerType("channel", [](const std::string& params) -> std::shared_ptr<::grpc::Channel> {
        std::vector<std::string> parts = factory::parseParameters(params, ",");
        if (parts.size() < 2) {
            throw std::invalid_argument(
                "Insecure channel requires at least 2 parameters: endpoint, channel_credentials, [call_credentials...]");
        }
        std::string endpoint = factory::normalizeArgument(parts[0], true);
        auto current_credentials = ChannelCredentialFactory::instance().create(parts[1]);
        for (size_t i = 2; i < parts.size(); ++i) {
            auto additional_credentials = CallCredentialFactory::instance().create(parts[i]);
            if (additional_credentials) {
                current_credentials = ::grpc::CompositeChannelCredentials(current_credentials, additional_credentials);
            } else {
                throw std::invalid_argument("Failed to create credentials for parameter: " + parts[i]);
            }
        }
        return ::grpc::CreateChannel(endpoint, current_credentials);
    });
}

auto registerCustomChannelType() -> void {
    ChannelFactory::instance().registerType("custom_channel", [](const std::string& params) -> std::shared_ptr<::grpc::Channel> {
        std::vector<std::string> parts = factory::parseParameters(params, ",");
        if (parts.size() < 3) {
            throw std::invalid_argument("Custom channel requires at least 3 parameters: endpoint, "
                                        "arguments_configuration, channel_credentials, [call_credentials...]");
        }
        std::string endpoint = factory::normalizeArgument(parts[0], true);
        std::string arguments_configuration = factory::normalizeArgument(parts[1], true);
        auto arguments = getChannelArguments(arguments_configuration);
        if (!arguments) {
            throw std::invalid_argument("Failed to get channel arguments for configuration: " + arguments_configuration);
        }
        auto current_credentials = ChannelCredentialFactory::instance().create(parts[2]);
        for (size_t i = 3; i < parts.size(); ++i) {
            auto additional_credentials = CallCredentialFactory::instance().create(parts[i]);
            if (additional_credentials) {
                current_credentials = ::grpc::CompositeChannelCredentials(current_credentials, additional_credentials);
            } else {
                throw std::invalid_argument("Failed to create credentials for parameter: " + parts[i]);
            }
        }
        auto channel = ::grpc::CreateCustomChannel(endpoint, current_credentials, arguments.value());
        return channel;
    });
}

auto registerInsecureCredentialsType() -> void {
    ChannelCredentialFactory::instance().registerType(
        "insecure", [](const std::string& /*params*/) -> std::shared_ptr<::grpc::ChannelCredentials> {
            return ::grpc::InsecureChannelCredentials();
        });
}

auto registerSslServerCredentialsType() -> void {
    ServerCredentialFactory::instance().registerType(
        "ssl_server", [](const std::string& params) -> std::shared_ptr<::grpc::ServerCredentials> {
            std::vector<std::string> parts = factory::parseParameters(params, ",");
            if (parts.size() != 2) {
                throw std::invalid_argument("SSL server credentials require exactly 2 parameters: pem_root_certs, private_key");
            }
            ::grpc::SslServerCredentialsOptions ssl_opts;
            ssl_opts.pem_key_cert_pairs.push_back({
                .private_key = filesystem::loadFile(factory::normalizeArgument(parts[1], true)), // private_key
                .cert_chain = filesystem::loadFile(factory::normalizeArgument(parts[0], true)),  // pem_root_certs
            });

            return ::grpc::SslServerCredentials(ssl_opts);
        });
}

auto registerSslClientCredentialsType() -> void {
    ChannelCredentialFactory::instance().registerType(
        "ssl_client", [](const std::string& params) -> std::shared_ptr<::grpc::ChannelCredentials> {
            std::vector<std::string> parts = factory::parseParameters(params, ",");
            if (parts.size() != 1) {
                throw std::invalid_argument("SSL client credentials require exactly 1 parameter: pem_root_certs");
            }
            ::grpc::SslCredentialsOptions ssl_opts;
            ssl_opts.pem_root_certs = filesystem::loadFile(factory::normalizeArgument(parts[0], true));
            return ::grpc::SslCredentials(ssl_opts);
        });
}

auto registerAuthCredentialsType() -> void {
    CallCredentialFactory::instance().registerType(
        AuthMetadataPlugin::REGISTERED_NAME, [](const std::string& params) -> std::shared_ptr<::grpc::CallCredentials> {
            std::vector<std::string> parts = factory::parseParameters(params, ",");
            if (parts.size() != 1) {
                throw std::invalid_argument("Auth credentials require exactly 1 parameter: auth_token");
            }
            std::string path_to_token = factory::normalizeArgument(parts[0], true);
            std::optional<std::string> token_opt = config::getTokenFromCredentials(path_to_token);
            if (token_opt == std::nullopt) {
                throw std::invalid_argument("Failed to retrieve token from credentials for parameter: " + path_to_token);
            }
            std::string token = token_opt.value();
            return ::grpc::MetadataCredentialsFromPlugin(std::make_unique<AuthMetadataPlugin>(token));
        });
}

auto registerAuthServerInterceptorFactoryType() -> void {
    ServerInterceptorFactoriesFactory::instance().registerType(
        AuthInterceptorFactory::REGISTERED_NAME,
        [](const std::string& params) -> std::unique_ptr<::grpc::experimental::ServerInterceptorFactoryInterface> {
            std::vector<std::string> parts = factory::parseParameters(params, ",");
            if (parts.size() != 1) {
                throw std::invalid_argument("Auth server interceptor factory requires exactly 1 parameter: expected_token");
            }
            std::string path_to_token = factory::normalizeArgument(parts[0], true);
            std::optional<std::string> token_opt = config::getTokenFromCredentials(path_to_token);
            if (token_opt == std::nullopt) {
                throw std::invalid_argument("Failed to retrieve token from credentials for parameter: " + path_to_token);
            }
            std::string expected_token = token_opt.value();
            return std::make_unique<AuthInterceptorFactory>(expected_token);
        });
}

} // namespace oink_judge::utils::grpc

File: app/utils/grpc/base_types.py
import grpc
import grpc.aio
from oink_judge.pybind11_utils_grpc import get_channel_arguments_list

from app.config import get_token_from_credentials
from app.factory import normalize_argument, parse_parameters
from app.utils import load_file

from .auth_metadata_plugin import TokenAuthPlugin
from .factories import CallCredentialFactory, ChannelCredentialFactory, ChannelFactory


def register_channel_type() -> None:
    def channel(params: str) -> grpc.aio.Channel:

        parts = parse_parameters(params, ",")
        if len(parts) < 2:
            raise ValueError(
                "Channel requires at least 2 parameters: endpoint, channel_credentials, [call_credentials...]"
            )
        endpoint = normalize_argument(parts[0], True)
        channel_credentials = ChannelCredentialFactory.instance().create(parts[1])
        for i in range(2, len(parts)):
            call_credentials = CallCredentialFactory.instance().create(parts[i])
            if call_credentials is not None:
                channel_credentials = grpc.composite_channel_credentials(
                    channel_credentials, call_credentials
                )
            else:
                raise ValueError(
                    f"Failed to create call credentials for parameter: {parts[i]}"
                )
        return grpc.aio.secure_channel(endpoint, channel_credentials)

    ChannelFactory.instance().register_type("channel", channel)


def register_custom_channel_type() -> None:
    def custom_channel(params: str) -> grpc.aio.Channel:
        parts = parse_parameters(params, ",")
        if len(parts) < 3:
            raise ValueError(
                "Custom channel requires at least 3 parameters: endpoint, arguments_configuration, channel_credentials, [call_credentials...]"
            )
        endpoint = normalize_argument(parts[0], True)
        arguments_configuration = normalize_argument(parts[1], True)
        arguments_list = get_channel_arguments_list(arguments_configuration)
        if arguments_list is None:
            raise ValueError(
                f"Failed to get channel arguments list for configuration: {arguments_configuration}"
            )

        channel_credentials = ChannelCredentialFactory.instance().create(parts[2])
        for i in range(3, len(parts)):
            call_credentials = CallCredentialFactory.instance().create(parts[i])
            if call_credentials is not None:
                channel_credentials = grpc.composite_channel_credentials(
                    channel_credentials, call_credentials
                )
            else:
                raise ValueError(
                    f"Failed to create call credentials for parameter: {parts[i]}"
                )
        return grpc.aio.secure_channel(endpoint, channel_credentials, arguments_list)

    ChannelFactory.instance().register_type("custom_channel", custom_channel)


def register_insecure_channel_type() -> None:
    def channel(params: str) -> grpc.aio.Channel:

        parts = parse_parameters(params, ",")
        if len(parts) != 1:
            raise ValueError("Insecure channel requires exactly 1 parameter: endpoint")
        endpoint = normalize_argument(parts[0], True)
        return grpc.aio.insecure_channel(endpoint)

    ChannelFactory.instance().register_type("insecure", channel)


def register_ssl_client_credentials_type() -> None:
    def ssl_client_credentials(params: str) -> grpc.ChannelCredentials:
        parts = parse_parameters(params, ",")
        if len(parts) != 1:
            raise ValueError(
                "SSL client credentials require exactly 1 parameter: path to pem root certs"
            )
        path_to_pem_root_certs = parts[0]
        pem_root_certs = load_file(path_to_pem_root_certs)
        return grpc.ssl_channel_credentials(pem_root_certs.encode())

    ChannelCredentialFactory.instance().register_type(
        "ssl_client", ssl_client_credentials
    )


def register_auth_credentials_type() -> None:
    def auth_credentials(params: str) -> grpc.CallCredentials:
        parts = parse_parameters(params, ",")
        if len(parts) != 1:
            raise ValueError(
                "Auth credentials require exactly 1 parameter: path to token"
            )
        path_to_token = parts[0]
        token = get_token_from_credentials(path_to_token)
        if token is None:
            raise ValueError(
                f"Failed to retrieve token from credentials for parameter: {path_to_token}"
            )
        auth_plugin = TokenAuthPlugin(token)
        return grpc.metadata_call_credentials(auth_plugin)

    CallCredentialFactory.instance().register_type("auth", auth_credentials)
